File: targets/python-flask/services/file_service.py
"""File retrieval logic."""
import os

import config

ALLOWED_FILES = {"readme.txt", "help.txt"}


def read_user_file(name):
    """Concatenates user-controlled name into a filesystem path (sink)."""
    path = os.path.join(config.UPLOAD_DIR, name)
    with open(path, "r") as f:
        return f.read()


def read_whitelisted(name):
    if name not in ALLOWED_FILES:
        raise ValueError("file not allowed")
    path = os.path.join(config.UPLOAD_DIR, name)
    with open(path, "r") as f:
        return f.read()

File: targets/csharp-aspnet/Services/FileService.cs
using System.Collections.Generic;
using System.IO;
using BadDemo.Config;

namespace BadDemo.Services
{
    /// <summary>File retrieval logic.</summary>
    public class FileService
    {
        private static readonly HashSet<string> Allowed = new HashSet<string> { "readme.txt", "help.txt" };

        /// <summary>Concatenates user-controlled name into a filesystem path (sink).</summary>
        public string ReadUserFile(string name)
        {
            var path = Path.Combine(AppConfig.UploadDir, name);
            return File.ReadAllText(path);
        }

        public string ReadWhitelisted(string name)
        {
            if (!Allowed.Contains(name))
            {
                throw new InvalidDataException("file not allowed");
            }
            return File.ReadAllText(Path.Combine(AppConfig.UploadDir, name));
        }
    }
}
